[4-9단계 - 방탈출 예약 관리] 체체(김진영) 미션 제출합니다.

CheChe903 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

하지만 그때 테스트는 하나도 작성하지 않았고, 제 마음대로 스프링을 사용하였습니다.


## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.


안녕하세요 오리! 이번에도 잘 부탁합니다.

제가 이번 단계에서 잘못 생각했던 점이 있습니다.

1. 내가 아는 스프링을 자꾸 쓰려 한다.
2. 하나를 써도 장단점을 파악하지 않고 쓴다.
3. 다른 방법도 있는데, 굳이 이 방법을 쓴 이유를 찾지 않았다.
4. 요구사항에 없는 것(예약은 미래만 가능하다.)에 집중했다.

이번에는 최대한 위 내용들을 배제하려 노력했습니다!


> 1레벨때를 떠올려봤을 때 콘솔의 입력값에 대한 검증보다는 도메인에 대해 집중했었는데 지금 구조는 그것이 아니라 그부분을 의식해보시란 이야기였어요.

이 부분에 대해서 기준이 잡히게 되었습니다.
도메인에 중요한 검증은 도메인에서 진행하고, null값, blank 값 검증은 DTO에서 진행합니다. 1단계 때 했던 방식과 똑같습니다.

> @RestController에서 이미 ok를 반환하도록 되어있는데 ResponseEntity를 또 래핑할 필요가 있나요? 즉, spring docs에서는 @RestController만으로도 ok를 반환하도록 만드는 것이 정의된 스펙인데 불필요한 객체 생성을 또 할 필요가 있을까요?

해당 부분에 있어 찾아보니 불필요하게 객체를 생성할 필요가 없을 것 같습니다.
기본적으로 200으로 제공하니, 200 응답은 객체 자체로 반환하고, 다른 상황(201 CREATED, DELETE 204)는 직접 정하여 보내주는 것이 맞다고 생각합니다!

> ReservationResponse에서 Reservation을 받아 생성하는 생성자를 만드는 것과 무슨 차이일까요?

'도메인이 DTO에 대해서 모른다.' 라고 생각합니다.  즉, 도메인이 DTO에 의존하지 않는다는 것입니다. 

이 부분이 중요하게 작용될 수 있을까요 ? 아직은 그러한 경험을 해보지 못하여 와닿지 않아요.

> 사용하지 않는 주석은 제거해주세요. bdd의 목적은 무조건 주석을 다는 것이 목적이 아니에요.

BDD의 목적이 무조건 주석을 다는 것이 아니라 다른 목적이라면, 주석의 유무는 상관이 없지않을까? 라는 생각도 했습니다.
단순히 제 관점에서는 비어있는 주석도 의도로 파악될 수 있다고 판단하여 두었어요.  (이 테스트는 선행 조건이 없어도 수행 가능한 테스트다.) 
오리는 어떠한 관점 때문에 주석이 없어도 된다고 생각하시는 지 궁금합니다!

CheChe903 committed:
feat(Clock): 비즈니스 로직에서 시간 의존성 제어 추가

기존에는 now 메서드를 사용하여 시간에 대한 제어를 할 수 없었는데, 이제는 Clock를 사용하여 기준 시간을 제어할 수 있도록 합니다.
RestAssured 테스트는 비즈니스 로직이 아닌, 통신 테스트에만 집중합니다.

CheChe903 committed:
refactor(ReservationController): 200 응답 객체 반환하도록 수정

spring docs에 따라 200 응답이 정해진 스펙이므로 객체만 반환하도록 합니다.

CheChe903 committed:
remove(ReservationController): 불필요한 메서드 삭제

CheChe903 committed:
refactor(Reservation): 동등성 검사 책임 Reservation 으로 이전

CheChe903 committed:
refactor: 입력값 검증 DTO 로 전환 및 @Valid 의존성 제거

기존에 객체에서 입력값이 null 또는 blank 를 검사하였지만 이젠 DTO에서 검사호도록 합니다.

CheChe903 committed:
feat: 4단계 구현

- [x] 데이터베이스 연결
- [x] 단계 테스트 추가

CheChe903 committed:
feat(Reservation): findAll 로직 DB로 연결

CheChe903 committed:
feat(Reservations): add, delete 로직 데이터베이스로 변환

CheChe903 committed:
refactor: id 타입 변환

확실한 값이 있는 id는 원시값인 long을 보냅니다.

CheChe903 committed:
feat(Reservations): 예약 일급 컬렉션 인터페이스화

CheChe903 committed:
refactor(JdbcReservations): rowMapper 수정

CheChe903 committed:
refactor(JdbcReservations): insert 쿼리 SimpleJdbcInsert이용하여 키 값 반환으로 수정

CheChe903 committed:
feat(ViewController): time 뷰 추가 및 미션 테스트 추가

CheChe903 committed:
feat(Time): 7단계 구현

CheChe903 committed:
refactor: 생성 요청 응답 201로 변경

CheChe903 committed:
style: 공백 컨벤션

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
chore(schema.sql): 스키마 수정

예약의 외래키를 예약 시간으로 설정합니다.

CheChe903 committed:
feat(ReservationTime): 예약 시간 로직 구현

CheChe903 committed:
feat(Reservation): 예약과 예약 시간 연동

CheChe903 committed:
refactor(MissionStepTest): 테스트 업데이트

CheChe903 committed:
test(ReservationTimeTest): 예약 시간 단위 테스트

CheChe903 committed:
refactor: 패키지 및 클래스 이름 변경

CheChe903 committed:
feat(ReservationService): 서비스 계층 추가

CheChe903 committed:
refactor: 웹에 맞게 수정

CheChe903 committed:
refactor(Reservation): 예약 생성 응답 DTO 사용

CheChe903 committed:
refactor(ReservationService): 계층 추상화 진행

CheChe903 committed:
test(ReservationService): 서비스 계층 테스트

CheChe903 committed:
test(ReservationTimeCreateRequest): DTO 단위 테스트

CheChe903 committed:
refactor: 패키지 변경

CheChe903 committed:
test(JdbcReservationTimeDaoTest): 단위 테스트 생성

CheChe903 committed:
test(JdbcReservationDaoTest): 단위 테스트 생성

CheChe903 committed:
refactor(ReservationServiceImplTest): 테스트 클래스명 변경

CheChe903 committed:
refactor(ReservationService): 추상화 삭제

CheChe903 committed:
refactor(ReservationServiceTest): 테스트 수정

CheChe903 committed:
docs(README): 요구사항, API 업데이트

jinyoungchoi95 commented on src/main/java/roomescape/dto/response/ReservationResponse.java:
```diff
@@ -1,7 +1,7 @@
+import roomescape.domain.ReservationTime;
 
-public record ReservationResponse(Long id, String name, LocalDate date, LocalTime time) {
+public record ReservationResponse(long id, String name, LocalDate date, ReservationTime time) {
```
> '도메인이 DTO에 대해서 모른다.' 라고 생각합니다. 즉, 도메인이 DTO에 의존하지 않는다는 것입니다.
> 이 부분이 중요하게 작용될 수 있을까요 ? 아직은 그러한 경험을 해보지 못하여 와닿지 않아요.

```java
public ReservationResponse(Reservation reservation) {
```
를 이야기한 것이고, 도메인이 dto에 대해서 모르는 것은 맞으나 지금 상황은 dto가 도메인을 가지고 생성하는 상황이기 때문에 말씀하시는 상황과 다릅니다.

CheChe903 replied:
캡슐화가 더 잘 지켜질 것 같습니다. 
생성하는 역할은 컨트롤러가 아닌 자기 자신이 갖고 있어야 하기 때문입니다. 

결과적으로 도메인 -> DTO 방식은 두 가지가 있습니다.

1. 생성자 파라미터로 객체를 넣어 DTO 생성
2. private로 생성자를 구현 후 객체를 받아 생성하는 정적 팩토리 메서드로 생성

1,2번 모두 캡슐화를 잘 지킬 수 있는 방법입니다.

1번은 구조적으로 더 깔끔하다는 장점이 있습니다.
2번은 생성자를 private으로 막아 dto의 생성 의도를 파악할 수 있게 합니다.

현재로서는 1번을 채택했지만, 2번도 나쁘지 않은 방법이라고 생각합니다. 팀 컨벤션에 따라 결정할 것 같아요.

그 반대인 DTO -> 도메인 비슷합니다.

현재는 DTO를 서비스 계층에서 찢어서 보냅니다. 하지만 이때 객체로 만들지 않고 필드를 하나씩 파라미터로 보냅니다.
이러한 경우는 만약 필드값이 바뀌었을 경우 서비스 계층에 영향이 갑니다. (서비스는 객체 자체만을 가지고 있다면 필드 변경 시 이에 대한 영향을 받지 않습니다.) 

결론
1. 컨트롤러 계층은 DTO로 관리
2. 서비스 계층은 받은 요청을 id가 null인 객체로 생성하여 repo 또는 dao에게 보냅니다. 보낼 요청은 객체를 DTO로 변환하여 보냅니다. 
3. 레포지토리 또는 dao 계층은 객체로만 관리합니다.


[refactor(DTO): DTO 사용 방식 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/cddb4b872ee1a0612ac826d8d9981a94a9fc2f52)

jinyoungchoi95 commented on src/test/java/roomescape/dto/request/ReservationTimeCreateRequestTest.java:
```diff
@@ -0,0 +1,22 @@
+    @Test
+    void validateTimeNullThrowExceptionTest() {
+
+        // given
```
> BDD의 목적이 무조건 주석을 다는 것이 아니라 다른 목적이라면, 주석의 유무는 상관이 없지않을까? 라는 생각도 했습니다.
> 단순히 제 관점에서는 비어있는 주석도 의도로 파악될 수 있다고 판단하여 두었어요. (이 테스트는 선행 조건이 없어도 수행 가능한 테스트다.)
> 오리는 어떠한 관점 때문에 주석이 없어도 된다고 생각하시는 지 궁금합니다!

선행 조건이 없어도 수행가능한 테스트라면 선행조건 자체가 코드에 없어도 되는 것 같아요. bdd의 목적은 상황을 나누기위한 목적이지 그상황이 발생하지 않더라도 주석으로 반드시 남겨두라가 목적이 아닙니다.

주석도 하나의 코드인데, 동작하지 않는 코드를 남겨두는 것이 맞는 행위일까요?

CheChe903 replied:
제 의견을 막 쓰다가, 오리에게 설득당한 것 같아요.

'선행 조건이 없어도 수행 가능한 테스트라면, 선행 조건 자체가 코드에 없어도 된다.' 

결국엔 given 절도 하나의 선행 코드이니, 작성하지 않는다면 이해하는데 큰 문제가 없을 것이라고 판단됩니다! 감사합니다.

[refactor(test): BDD 필요없는 주석 제거](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/30ffac570bb56418a8ad26b42366b0e082053565)

jinyoungchoi95 commented on src/main/java/roomescape/controller/TimeController.java:
```diff
@@ -0,0 +1,52 @@
+        ReservationTimeCreateResponse reservationTimeCreateResponse = new ReservationTimeCreateResponse(
+                reservationTimeDao.create(reservationTimeCreateRequest),
+                reservationTimeCreateRequest.startAt());
+        return ResponseEntity.status(HttpStatus.CREATED).body(reservationTimeCreateResponse);
```
https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/201

created의 반환값에는 어떤 것을 기대하는지, `ResponseEntity.created` 메서드는 어떤 것을 생성하도록 유도하는지 학습해보면 좋겠네요.

CheChe903 replied:
해당 페이지에서는 응답에 uri도 함께 포함하고 있습니다. 하지만 uri를 포함하는 이유에 대해 궁금하여 찾아보았습니다.

>하나의 데이터를 생성한 후 데이터의 상세정보를 볼 때 id값을 다시한번 서버에 물어봐야 한다. 하지만 해당 URI를 반환시킨다면, 다시 물어보는 만큼의 네트워크 트래픽을 줄일 수 있기 때문에 사용하는 것이다.

하지만, 궁금증이 하나 더 생겼습니다. 바디에도 id를 넣는데, 왜 굳이 uri로 한 번 더 넣어줄까 ?

1. 백엔드 uri가 바뀌어도 프론트엔드가 모를 수 있다.
  이는 적합한 이유로 다가오지 않았습니다. uri가 바뀌는 것은 중요한 변경이고, 이는 이미 협의된 상태에서 바뀌어야 한다고 생각합니다. 
    1-1. 추가적으로, 로케이션을 제공함으로써 프론트엔드가 그것만으로 기본 uri를 만든다.
      하지만 이미 만들어진 리소스들을 이용하여 추후에 쓴다면 uri 정보가 사라진다면 오류가 발생할 수 있지 않을까 생각됩니다.

2. swagger, postman과 같은 테스트 도구에 사용된다.
  이 이유는 표준이 location을 넣기 때문에 테스트 도구에 이것이 이용된 것이라고 생각됩니다.

 현재로서의 결론은 '표준에 따른 명시적 커뮤니케이션' 이 가장 적합한 이유로 생각됩니다. 
 이 이유로는 아직 해야 할 필요성을 찾지 못하였어요. 
 
 혹시 오리가 학습할 수 있는 관련 키워드를 줄 수 있을까요 ?


먼저, 학습을 위해 그렇게 수정하였습니다.


[feat(Controller): CREATED 헤더에 URI 추가](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/df77e6c1e45ff475d3599b77ba27668fab92d617)

jinyoungchoi95 replied:
정의된 http 스펙이기 때문이에요. 바디를 어떻게 쓸지, uri를 어떻게 쓸지는 개발자가 선택하는 영역입니다. 현재 요청한 생성완료로 인한 바디에서는 생성완료로 인한 간단한 정보만 표기하고 이 생성완료된 상세 정보는 어떤 uri를 통해서 확인할 수 있는지에 대한 정보를 나타내는거죠.

관련된 키워드드들은 네트워크에 대한 학습이 필요한 내용으로 lms에 학습을 위한 참고도서들이 있으니 해당 도서들을 참고해보면 좋겠네요.

jinyoungchoi95 commented on src/main/java/roomescape/controller/TimeController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @GetMapping
+    public List<ReservationTimeResponse> findAll() {
+        return reservationTimeDao.findAll().stream()
```
service를 도입한 곳과 하지 않은 곳의 차이는 무엇인가요?

CheChe903 replied:
현재 서비스를 도입한 Reservation은 ReservationTime이 필요하기에, 컨트롤러에서 이를 해결한다면 책임이 커집니다.

단순한 crud는 컨트롤러 - dao 에서 해결할 수 있지만, Time의 dao가 필요한 이상 서비스 계층에서 처리하는 것이 책임을 올바르게 배정했다고 생각되어 서비스 계층을 추가하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/dao/reservation/InMemoryReservationDao.java:
어디에 사용되나요? main에 있어도 되는 클래스인가요?

CheChe903 replied:
서비스 단위 테스트에서 사용되는 파일입니다. 테스트로 이동하였습니다.

[refactor: 페이크 객체 패키지 이동](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/469ae56911b0d78bc538ee30d01c5a32347aea41)

jinyoungchoi95 commented on src/main/java/roomescape/dao/reservation/JdbcReservationDao.java:
```diff
@@ -0,0 +1,73 @@
+                .withTableName("reservation")
+                .usingGeneratedKeyColumns("id");
+
+        Map<String, Object> parameters = new HashMap<>();
```
`Map.of()`를 써볼 수 있겠네요.

CheChe903 replied:
그렇네요, 레벨1을 망각하지 않도록 노력하겠습니다.

[refactor(Dao): SimpleJdbcInsert 메서드 중 Map의 of 메서드 사용](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/dcb71c502e82821d618c338c744bab23109a5568)

jinyoungchoi95 commented on src/main/java/roomescape/dao/reservation/ReservationDao.java:
```diff
@@ -0,0 +1,14 @@
+
+    long create(ReservationCreateRequest reservationCreateRequest);
+
+    void delete(final Long id);
```
```suggestion
    void delete(long id);
```
1. 인터페이스에는 final이 없어도 되겠네요.
2. id를 무조건 받아야할건데 nullable long을 써도 될까요?

CheChe903 replied:
2번에 있어서 학습하였습니다.

원시값 - null을 고려해야 하는 상황, 즉 요청 바디 DTO처럼 값이 안들어올 수 있는 경우, id 같은 PK에 없음을 표시하기 위해 (원시는 null일 경우 0으로 계산)

객체값 - null을 고려하지 않아도 되는 상황, 즉 값이 무조건 들어오는 경우 PathVariable은 값이 어떠든 결국 uri에 표시되어야 요청이 들어오기 때문에 null일 필요 없다. findById, Response 등 특이 사항이 없는 경우


[refactor: null을 고려해야하는 상황 객체로 포장하여 사용](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/09324e5330ef3f3d391cf9c98ea3cb5e5f33cdab)

jinyoungchoi95 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,16 @@
+
+CREATE TABLE reservation
+(
+    id      BIGINT       NOT NULL AUTO_INCREMENT,
+    name    VARCHAR(255) NOT NULL,
+    date    VARCHAR(255) NOT NULL,
```
db에는 글자수 제한이 있는데 도메인에는 없네요.

CheChe903 replied:
name은 쉽게 검증해줄 수 있었지만, LocalTime, LocalDate는 그렇게 들어와 메핑되어 제어할 수 없는 부분이였습니다. 즉, 검증을 제가 제어할 수 없었습니다.
그래서 제어권을 받고자 request를 String으로 변경 후 검증 작업을 진행했습니다. 더 큰 범위인 format 검증으로 255자보다 더 강하게 검증하였습니다.

[feat: Database 제약 조건에 맞는 검증 추가](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/f9b747eb518bb8a8a36b90167aef51957d9cec63)

jinyoungchoi95 replied:
name에 해당하는 코멘트였어요. date는 db에 날짜를 표기하는 타입을 학습해보시면 좋겠네요.

jinyoungchoi95 commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -1,37 +1,23 @@
         this.time = time;
     }
 
-    private void validateName(final String name) {
-        if (name == null || name.isBlank()) {
-            throw new IllegalArgumentException("이름은 비어있을 수 없습니다.");
-        }
-    }
-
-    private void validateDateTime(final LocalDate date, final LocalTime time) {
-        if (date.isBefore(LocalDate.now())) {
-            throw new IllegalArgumentException("예약 시간은 과거일 수 없습니다.");
-        }
-        if (date.isEqual(LocalDate.now()) && time.isBefore(LocalTime.now())) {
-            throw new IllegalArgumentException("예약 시간은 과거일 수 없습니다.");
-        }
```
지워진 이유가 무엇일까요?

CheChe903 replied:
현재 없는 요구사항입니다. 이름 검증은 DTO로 넘겼습니다.
미래 시간에 대한 요구사항은, 이를 신경쓰며 현재 요구사항에 집중 못하여 삭제하였습니다.

jinyoungchoi95 commented on src/test/java/roomescape/dao/reservation/JdbcReservationDaoTest.java:
```diff
@@ -0,0 +1,86 @@
+import roomescape.dto.request.ReservationTimeCreateRequest;
+
+@SpringBootTest
+@TestPropertySource(locations = "classpath:application-test.properties")
```
지우고 `application-test.properties`를 `application.properties`로 하면 동일하게 실행되지 않을까요?

jinyoungchoi95 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,21 +1,43 @@
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
 public class MissionStepTest {
 
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    private Connection connection;
+
+    @BeforeEach
+    void setUp() throws SQLException {
+        connection = jdbcTemplate.getDataSource().getConnection();
+    }
+
+    @AfterEach
+    void tearDown() throws SQLException {
+        if (connection != null && !connection.isClosed()) {
+            connection.close();
+        }
+    }
```
어디에 사용되고 있는 코드일까요?

CheChe903 replied:
JdbcTemplate를 이용하기 위한 Connection을 얻고, 테스트 하기 위해 사용 되었습니다.

이를 사용되지 않는 테스트도 존재하지만, MissionTest라 두었습니다. 중첩 클래스로 구분짓는 것이 좋을까요 ?

jinyoungchoi95 replied:
connection을 모든 테스트에서 사용하지 않은데 전역적으로 설정할 필요는 없는 것 같아요. connection 정보를 테스트하고싶다면 특정 테스트 메서드에서 수행하면 되지 않을까요?

jinyoungchoi95 commented on src/test/java/roomescape/domain/ReservationTest.java:
```diff
@@ -1,85 +1,27 @@
-    void validateNameNullThrowExceptionTest() {
+    @DisplayName("아이디가 같으면 true를, 다르면 false를 반환한다.")
+    @ParameterizedTest
+    @CsvSource(value = "1, 1, true, 1, 2, false")
```
다 `,`로 구분지어있으면 어디까지가 데이터 구분선인지 명확하게 눈에 안보일 것 같네요. value list로 구분지으면 되지 않을까요?

jinyoungchoi95 commented on src/test/java/roomescape/dao/reservation/JdbcReservationDaoTest.java:
```diff
@@ -0,0 +1,86 @@
+    @Autowired
+    private JdbcReservationTimeDao jdbcReservationTimeDao;
+
+    @BeforeEach
+    void setUp() {
+        jdbcTemplate.update("DELETE FROM reservation");
+        jdbcTemplate.update("DELETE FROM reservation_time");
+    }
```
`@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)`를 사용하지 않은 이유가 있을까요?

CheChe903 replied:
두가지 방법을 찾았습니다.

1. 더티 컨텍스트로 스프링에게 재사용하지 말라고 알리는 방법
이는 스프링 컨테이너 전체를 띄워야 하기 때문에 리소스가 많이 듭니다.

2. @JdbcTest를 이용한 Transactional 어노테이션 사용
@JdbcTest로 필요한 빈만 명시적으로 사용할 수 있습니다. 이렇게 되면 사용하지 않는 컨텍스트들은 컨테이너에 올라가지 않습니다.

추가적으로 @JdbcTest 어노테이션은 @Transactional 을 갖고 있습니다. 기본적으로 테스트 환경에서는 rollback = true가 디폴트 값으로 사용되어 현재 테스트에서는 더티 컨텍스트와 똑같은 기능을 합니다.

그럼에도 비용이 더 큰 더티컨텍스트를 사용해야하는 상황을 찾아보았습니다.

1. 스키마 변경은 트랜잭션으로 롤백되지 않음
2. 캐시 상태는 트랜잭션으로 롤백되지 않음

현재까지는 두 상황만 이해가 됩니다. 

[test(Jdbc): Jdbc 테스트 어노테이션 변경 및 application.properties 파일 이름 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/299/commits/87ac44825e4d058dbb5f2578812bd8b499ac386b)

jinyoungchoi95 replied:
스프링 컨테이너를 이야기 주셨는데 해당 내용은 다음 단계 미션을 진행하실 때에 학습하게 될 내용이 될 것 같아 넘어가겠습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 체체 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

CheChe903 committed:
refactor(DTO): DTO 사용 방식 변경

1. DTO는 스스로가 객체를 통해 생성하도록 합니다.
2. DTO는 컨트롤러 - 서비스 요청 및 컨트롤러의 응답에만 쓰입니다. 서비스 계층은 이를 객체로 변환하고, Dao에서는 객체를 받아 관리합니다.
3. 이를 위해 전반적인 코드를 수정합니다.

CheChe903 committed:
refactor(test): BDD 필요없는 주석 제거

CheChe903 committed:
feat(Controller): CREATED 헤더에 URI 추가

CheChe903 committed:
refactor(FakeDao): 페이크 객체 test 패키지로 이동

CheChe903 committed:
refactor(Dao): SimpleJdbcInsert 메서드 중 Map의 of 메서드 사용

CheChe903 committed:
refactor: null을 고려해야하는 상황 객체로 포장하여 사용

CheChe903 committed:
feat: Database 제약 조건에 맞는 검증 추가

1. 이름은 255자 이하여야 합니다.
2. 날짜, 시간도 255자 이하지만, 이는 날짜, 시간 형식 포맷의 체크로 검증합니다.

CheChe903 committed:
refactor: 페이크 객체 패키지 이동

CheChe903 committed:
test(Jdbc): Jdbc 테스트 어노테이션 변경 및 application.properties 파일 이름 변경

CheChe903 committed:
test(ReservationTest): 이름 길이 검증 테스트 추가

jinyoungchoi95 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -11,54 +10,43 @@
+            @RequestBody ReservationCreateRequest reservationCreateRequest) {
+        ReservationCreateResponse reservationCreateResponse = reservationService.create(reservationCreateRequest);
+
+        URI location = ServletUriComponentsBuilder
+                .fromCurrentRequest()
+                .path("/{id}")
+                .buildAndExpand(reservationCreateResponse.getId())
+                .toUri();
```
그래서 드렸던 코멘트는 201이라는 스펙에 대해서 학습해보는 것을 말씀드리는 내용이었고 현재 location은 사용자가 쓸 수 있도록 스펙 제공을 하지 않으니 해당 api를 사용하는 입장에서는 버그가 발생할 수 있으므로 location은 제거되어야겠네요.

jinyoungchoi95 reviewed (approved):
안녕하세요 체체 😄

현재 미션에서 필요한 학습을 모두 하신 것 같아 머지하겠습니다.
방탈출 예약 관리 미션기간동안 주고받았던 피드백이 앞으로도 많은 도움이 되길 바랍니다 😄

감사합니다 🙇‍♂️
